source/grid/manifold.cc

```diff
@@ -157,85 +157,54 @@ normal_vector (const Triangulation<3, 3>::face_iterator &face,
 {
   const int spacedim=3;
   Tensor<1,spacedim> t1,t2;
+  Tensor<1,spacedim> normal;
 
-  // Take the difference between p and all four vertices
-  unsigned int min_index=0;
-  double       min_distance = (p-face->vertex(0)).norm_square();
-
-  for (unsigned int i=1; i<4; ++i)
-    {
-      const Tensor<1,spacedim> dp = p-face->vertex(i);
-      double distance = dp.norm_square();
-      if (distance < min_distance)
-        {
-          min_index = i;
-          min_distance = distance;
-        }
-    }
-  // Verify we have a valid vertex index
-  AssertIndexRange(min_index, 4);
-
-  // Now figure out which vertices are best to compute tangent vectors.
-  // We split the cell in a central diamond of points closer to the
-  // center than to any of the vertices, and the 4 triangles in the
-  // corner. The central diamond is split into its upper and lower
-  // half. For each of these 6 cases, the following encodes a list
-  // of two vertices each to which we compute the tangent vectors,
-  // and then take the cross product. See the documentation of this
-  // function for exact details.
-  if ((p-face->center()).norm_square() < min_distance)
+  // Look for a combination of tangent vectors that
+  // are of approximately equal length and not linearly dependent
+  for (unsigned int i=0,j=1 ; i<4 && j<4; ++j)
     {
-      // we are close to the face center: pick two consecutive vertices,
-      // but not the closest one. We make sure the direction is always
-      // the same.
-      if (min_index < 2)
-        {
-          t1 = get_tangent_vector(p, face->vertex(3));
-          t2 = get_tangent_vector(p, face->vertex(2));
-        }
-      else
+      // if p is too close to vertex i try again with different i and j
+      if ((p - face->vertex(i)).norm_square() <
+          std::numeric_limits<double>::epsilon() * (p - face->vertex(j)).norm_square())
```

Comment:
I think you should err on the safe side and make this `100 * eps` or some such. In fact, since you're comparing squares, you may even want to use `10000 * eps`.